src/idpi/grib_decoder.py

```diff
@@ -82,6 +52,171 @@ def _extract_pv(pv):
+        self.values[key] = field.to_numpy(dtype=np.float32)
+
+        step = key[-2]  # assume all members share the same time steps
+        if step not in self.time_meta:
+            self.time_meta[step] = field.metadata(namespace="time")
+
+        if not self.dims:
+            self.dims = tuple(DIM_MAP[d] for d in dim_keys) + ("y", "x")
+
+        if not self.metadata:
+            self.metadata = {
+                "message": field.message(),
+                **metadata.extract(field.metadata()),
+            }
+
+        if not self.hcoords:
+            self.hcoords = {
+                dim: xr.DataArray(dims=("y", "x"), data=values)
+                for dim, values in field.to_latlon().items()
+            }
+
+    def _gather_coords(self):
+        if self.dims is None:
+            raise RuntimeError("No dims.")
+
+        coord_values = zip(*self.values)
+        unique = (sorted(set(values)) for values in coord_values)
+        coords = {dim: c for dim, c in zip(self.dims[:-2], unique)}
+
+        if missing := [
+            combination
+            for combination in product(*coords.values())
+            if combination not in self.values
+        ]:
+            msg = f"Missing combinations: {missing}"
+            logger.exception(msg)
+            raise RuntimeError(msg)
+
+        ny, nx = next(iter(self.values.values())).shape
+        shape = tuple(len(v) for v in coords.values()) + (ny, nx)
+        return coords, shape
+
+    def _gather_tcoords(self):
+        time = None
+        valid_time = []
+        for step in sorted(self.time_meta):
+            tm = self.time_meta[step]
+            valid_time.append(_parse_datetime(tm["validityDate"], tm["validityTime"]))
+            if time is None:
+                time = _parse_datetime(tm["dataDate"], tm["dataTime"])
+
+        return {"valid_time": ("time", valid_time), "ref_time": time}
+
+    def to_xarray(self) -> xr.DataArray:
+        if not self.values:
+            raise RuntimeError("No values.")
+
+        coords, shape = self._gather_coords()
+        tcoords = self._gather_tcoords()
+
+        array = xr.DataArray(
+            data=np.array(
+                [self.values.pop(key) for key in sorted(self.values)]
+            ).reshape(shape),
+            coords=coords | self.hcoords | tcoords,
+            dims=self.dims,
+            attrs=self.metadata,
+        )
+
+        if array.vcoord_type != "surface":
+            return array
+
+        return array.squeeze("z", drop=True)
+
+
+def _load_buffer_map(
+    source: data_source.DataSource,
+    request: Request,
+) -> dict[str, _FieldBuffer]:
+    logger.info("Retrieving request: %s", request)
+    fs = source.retrieve(request)
+
+    buffer_map: dict[str, _FieldBuffer] = {}
+
+    for field in fs:
+        name = field.metadata("shortName")
+        buffer = buffer_map.setdefault(name, _FieldBuffer())
+        buffer.load(field)
+
+    return buffer_map
+
+
+def load_single_param(
```

Comment:
Would here a check make sense to verify that the request only contains a single param?

Reply:
Yes I think it would. I'll add it